In [1]:
import pandas as pd

In [2]:
train = pd.read_csv("data/train.csv")
test = pd.read_csv("data/test.csv")

In [3]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

We have following 10 features of each passenger:

* *Pclass* is the Ticket-class: first(1), second(2), and third(3) class tickets. This is an ordinal integer feature.

* *Name* is the name of passenger. Although the name itself does not reveal too much information to prediction of survival, there is some titles attached to names that can indicate a certain group. For example, Mrs is an indication of marriage. This is a categorical text string feature.

* *Sex* is the gender of passenger. Either female or male. This is a categorical text string feature.

* *Age* is the integer age of passenger. There might be some NaN values in this column. This is an integer feature.

* *SibSp* is the number of siblings and spouse aboard the Titanic. Sibling includes brother, sister, stepbrother, and stepsister, while spouse includes husband or wife (mistresses and fiances were ignored). This is an ordinal integer feature.

* *Parch* is the number of parents and children aboard the Titanic. This is also an ordinal integer feature.

* *Ticket* is a character string that gives the ticket number.

* *Fare* is a float feature showing how much money the passenger paid for their trip.

* *Cabin* is the cabin number of each passenger. There are NaN in this column. This is also another string feature.

* *Embarked* is the port of embarkation as a categorical character feature.

In summary we have 1 floating point feature (fare), 1 integer variable (age), 3 ordinal integer features (plcass, sibsp, parch), 2 categorical features (sex, embarked), and 3 text string features (ticket, cabin, name).

In [4]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Missing values

Take a look at missing values in training data:

In [5]:
print(train.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


It can be seen that 177 values were missing in Age, 687 values were missing in Cabin, and 2 values were missing in Embarked.

Let's fill in the empty values with the available information. Firstly, let's take a look at missing values of Embarked:

In [6]:
df = train[train["Embarked"].isnull()]
df

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

Then we find similar passengers:

In [7]:
train[(train["Sex"]=="female") & (train["SibSp"]==0) & (train["Parch"]==0) & (train["Fare"]>70.0) & (train["Fare"]<90.0)]

PassengerId  Survived  Pclass  \
61            62         1       1   
218          219         1       1   
256          257         1       1   
257          258         1       1   
290          291         1       1   
310          311         1       1   
504          505         1       1   
627          628         1       1   
759          760         1       1   
829          830         1       1   

                                                  Name     Sex   Age  SibSp  \
61                                 Icard, Miss. Amelie  female  38.0      0   
218                              Bazzani, Miss. Albina  female  32.0      0   
256                     Thorne, Mrs. Gertrude Maybelle  female   NaN      0   
257                               Cherry, Miss. Gladys  female  30.0      0   
290                       Barber, Miss. Ellen "Nellie"  female  26.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
504                              Maioni, Miss. Roberta  female  16.0      0   
627                      Longley, Miss. Gretchen Fiske  female  21.0      0   
759  Rothes, the Countess. of (Lucy Noel Martha Dye...  female  33.0      0   
829          Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   

     Parch    Ticket     Fare Cabin Embarked  
61       0    113572  80.0000   B28      NaN  
218      0     11813  76.2917   D15        C  
256      0  PC 17585  79.2000   NaN        C  
257      0    110152  86.5000   B77        S  
290      0     19877  78.8500   NaN        S  
310      0     11767  83.1583   C54        C  
504      0    110152  86.5000   B79        S  
627      0     13502  77.9583    D9        S  
759      0    110152  86.5000   B77        S  
829      0    113572  80.0000   B28      NaN

We can see that the similar passengers either embarked from C or S. Let's make it S here

In [8]:
train.ix[61,"Embarked"] = "S"
train.ix[829,"Embarked"] = "S"

Confirm there is no null values in column of Embarked now:

In [9]:
print(train.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64


In [10]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

There is a good explanation of [feature extraction and derivation](https://www.kaggle.com/headsortails/pytanic) in Kaggle. But to make things as simple as possible, let's just simply select the following features: Pclass, Sex, Age, SibSp, ParCh, Fare, Embarked.

## Preparing the data

Since not all classifiers can handle string input, it's a better practice to adjust the column types to numbers(integers or floats). For instance, string categorical feature Sex can be represent by 1(male) and 0(female).

In [11]:
print(train.Pclass.unique())
print(train.Sex.unique())
print(train.SibSp.unique())
print(train.Parch.unique())
print(train.Embarked.unique())

[3 1 2]
['male' 'female']
[1 0 3 4 2 5 8]
[0 1 2 5 3 4 6]
['S' 'C' 'Q']


In [12]:
train["Sex"] = train["Sex"].astype("category")
train["Sex"].cat.categories = [0,1]
train["Sex"] = train["Sex"].astype("int")

train["Embarked"] = train["Embarked"].astype("category")
train["Embarked"].cat.categories = [0,1,2]
train["Embarked"] = train["Embarked"].astype("int")

In [13]:
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Cabin  Embarked  
0         A/5 21171   7.2500   NaN         2  
1          PC 17599  71.2833   C85         0  
2  STON/O2. 3101282   7.9250   NaN         2  
3            113803  53.1000  C123         2  
4            373450   8.0500   NaN         2

## Modeling

### Splitting the train sample into two sub-samples: training and testing

To avoid overfitting, it's best practice to split the training data further into two sets.

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
training, testing = train_test_split(train, test_size=0.2, random_state=0)
print("Total sample size = %i; training sample size = %i, testing sample size = %i"\
     %(train.shape[0],training.shape[0],testing.shape[0]))

Total sample size = 891; training sample size = 712, testing sample size = 179


### Look at model features

In [16]:
cols = ["Pclass","Sex","Age","SibSp","Parch","Fare","Embarked"]
tcols = ["Survived"] + cols

In [17]:
df = training[tcols].dropna()

In [18]:
X = df.loc[:,cols]

In [19]:
y = df.loc[:,['Survived']]
y = [val[0] for val in y.values]

In [20]:
from sklearn.linear_model import LogisticRegression

In [21]:
clf_log = LogisticRegression()
clf_log = clf_log.fit(X,y)
score_log = clf_log.score(X,y)
print(score_log)

0.784588441331


In [22]:
coef = clf_log.coef_[0]

In [23]:
pd.DataFrame(list(zip(cols,coef)))

0         1
0    Pclass -0.825371
1       Sex -2.233182
2       Age -0.032287
3     SibSp -0.312647
4     Parch -0.021213
5      Fare  0.004752
6  Embarked  0.010840

TODO: negative meanings?

### Model validation

In [24]:
from sklearn.model_selection import cross_val_score
import numpy as np

In [25]:
scores = cross_val_score(clf_log, X, y, cv=5)
print(scores)
print("Mean score = %.3f, Std deviation = %.3f"%(np.mean(scores),np.std(scores)))

[ 0.74782609  0.8245614   0.78070175  0.78947368  0.78947368]
Mean score = 0.786, Std deviation = 0.024


Final validation with the testing data set:

In [26]:
df_test = testing[tcols].dropna()
X_test = df_test.loc[:,cols]
y_test = df_test.loc[:,['Survived']]
y_test = [val[0] for val in y_test.values]
score_log_test = clf_log.score(X_test,y_test)  # clf_log has been trained by training data
print(score_log_test)

0.79020979021


## Preparing prediction for submission

Finally, we use our model to predict the test data, and write to a submission file according to the submission rules (418 rows; only include the columns PassengerId and Survived).

In [27]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

Firstly, prepare the data for prediction.

In [28]:
test["Sex"] = test["Sex"].astype("category")
test["Sex"].cat.categories = [0,1]
test["Sex"] = test["Sex"].astype("int")

test["Embarked"] = test["Embarked"].astype("category")
test["Embarked"].cat.categories = [0,1,2]
test["Embarked"] = test["Embarked"].astype("int")

In [29]:
test.head()

PassengerId  Pclass                                          Name  Sex  \
0          892       3                              Kelly, Mr. James    1   
1          893       3              Wilkes, Mrs. James (Ellen Needs)    0   
2          894       2                     Myles, Mr. Thomas Francis    1   
3          895       3                              Wirz, Mr. Albert    1   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)    0   

    Age  SibSp  Parch   Ticket     Fare Cabin  Embarked  
0  34.5      0      0   330911   7.8292   NaN         1  
1  47.0      1      0   363272   7.0000   NaN         2  
2  62.0      0      0   240276   9.6875   NaN         1  
3  27.0      0      0   315154   8.6625   NaN         2  
4  22.0      1      1  3101298  12.2875   NaN         2

In [30]:
df_2 = test.loc[:,cols]

Firstly, we need to deal with missing values in test data.

In [31]:
print(df_2.isnull().sum())

Pclass       0
Sex          0
Age         86
SibSp        0
Parch        0
Fare         1
Embarked     0
dtype: int64


As we can see, 86 missing values in Age. Since Age is one of our selected features, we need to fillin the blanks with some default values.

In [32]:
df_2 = df_2.fillna(method="pad")

In [33]:
print(df_2.isnull().sum())

Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64


Now the test data is ready for prediction:

In [34]:
surv_pred = clf_log.predict(df_2)

In [35]:
surv_pred

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 0,

In [36]:
submit = pd.DataFrame({"PassengerID": test["PassengerId"],
                       "Survived": surv_pred})

In [37]:
submit

PassengerID  Survived
0            892         0
1            893         0
2            894         0
3            895         0
4            896         1
5            897         0
6            898         1
7            899         0
8            900         1
9            901         0
10           902         0
11           903         0
12           904         1
13           905         0
14           906         1
15           907         1
16           908         0
17           909         0
18           910         1
19           911         0
20           912         0
21           913         0
22           914         1
23           915         1
24           916         1
25           917         0
26           918         1
27           919         0
28           920         0
29           921         0
..           ...       ...
388         1280         0
389         1281         0
390         1282         1
391         1283         1
392         1284         0
393         1285         0
394         1286         0
395         1287         1
396         1288         0
397         1289         1
398         1290         0
399         1291         0
400         1292         1
401         1293         0
402         1294         1
403         1295         1
404         1296         0
405         1297         0
406         1298         0
407         1299         0
408         1300         0
409         1301         1
410         1302         1
411         1303         1
412         1304         1
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [38]:
submit.shape

(418, 2)

In [39]:
submit.to_csv("out.csv",index=False)